python/cuda_parallel/cuda/parallel/experimental/algorithms/unique_by_key.py

```diff
@@ -0,0 +1,207 @@
+# Copyright (c) 2024-2025, NVIDIA CORPORATION & AFFILIATES. ALL RIGHTS RESERVED.
+#
+#
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+import ctypes
+from typing import Callable
+
+import numba
+from numba import cuda
+from numba.cuda.cudadrv import enums
+
+from .. import _cccl as cccl
+from .._bindings import get_bindings, get_paths
+from .._caching import CachableFunction, cache_with_key
+from .._utils import protocols
+from ..iterators._iterators import IteratorBase
+from ..typing import DeviceArrayLike
+
+
+def make_cache_key(
+    d_in_keys: DeviceArrayLike | IteratorBase,
+    d_in_items: DeviceArrayLike | IteratorBase,
+    d_out_keys: DeviceArrayLike | IteratorBase,
+    d_out_items: DeviceArrayLike | IteratorBase,
+    d_out_num_selected: DeviceArrayLike,
+    op: Callable,
+):
+    d_in_keys_key = (
+        d_in_keys.kind
+        if isinstance(d_in_keys, IteratorBase)
+        else protocols.get_dtype(d_in_keys)
+    )
+    d_in_items_key = (
+        d_in_items.kind
+        if isinstance(d_in_items, IteratorBase)
+        else protocols.get_dtype(d_in_items)
+    )
+    d_out_keys_key = (
+        d_out_keys.kind
+        if isinstance(d_out_keys, IteratorBase)
+        else protocols.get_dtype(d_out_keys)
+    )
+    d_out_items_key = (
+        d_out_items.kind
+        if isinstance(d_out_items, IteratorBase)
+        else protocols.get_dtype(d_out_items)
+    )
+    d_out_num_selected_key = protocols.get_dtype(d_out_num_selected)
+    op_key = CachableFunction(op)
+
+    return (
+        d_in_keys_key,
+        d_in_items_key,
+        d_out_keys_key,
+        d_out_items_key,
+        d_out_num_selected_key,
+        op_key,
+    )
+
+
+def _update_device_array_pointers(current_array, passed_array):
+    if current_array.type.value == cccl.IteratorKind.POINTER:
+        current_array.state = protocols.get_data_pointer(passed_array)
+    else:
+        current_array.state = passed_array.state
+
+
+class _UniqueByKey:
+    def __init__(
+        self,
+        d_in_keys: DeviceArrayLike | IteratorBase,
+        d_in_items: DeviceArrayLike | IteratorBase,
+        d_out_keys: DeviceArrayLike | IteratorBase,
+        d_out_items: DeviceArrayLike | IteratorBase,
+        d_out_num_selected: DeviceArrayLike,
+        op: Callable,
+    ):
+        # Referenced from __del__:
+        self.build_result = None
+
+        self.d_in_keys_cccl = cccl.to_cccl_iter(d_in_keys)
+        self.d_in_items_cccl = cccl.to_cccl_iter(d_in_items)
+        self.d_out_keys_cccl = cccl.to_cccl_iter(d_out_keys)
+        self.d_out_items_cccl = cccl.to_cccl_iter(d_out_items)
+        self.d_out_num_selected_cccl = cccl.to_cccl_iter(d_out_num_selected)
+
+        cc_major, cc_minor = cuda.get_current_device().compute_capability
+        cub_path, thrust_path, libcudacxx_path, cuda_include_path = get_paths()
+        bindings = get_bindings()
```

Comment:
Let's save `bindings` in `self.bindings` and reuse in `__call__` and ``__del__`` special methods.